quickwit-index-config/src/tag_pruning.rs

```diff
@@ -0,0 +1,223 @@
+        TagFiltersAST::Or(conditions) if conditions.len() > 1 => {
+            let pruned_conditions: Vec<TagFiltersAST> = conditions
+                .into_iter()
+                .flat_map(|c| simplify_ast(c))
+                .collect();
+            Some(TagFiltersAST::Or(pruned_conditions))
+        }
+        ast @ TagFiltersAST::Tag { tag: _ } => Some(ast),
+        TagFiltersAST::Not(ast) => simplify_ast(*ast).map(|c| TagFiltersAST::Not(Box::new(c))),
+        TagFiltersAST::NoTagFound => None,
+        _ => None,
+    }
+}
+fn prune_unusable_filters(filter: TagFiltersAST) -> Option<TagFiltersAST> {
+    match filter {
+        TagFiltersAST::And(conditions) if conditions.len() == 1 => {
+            let cond = conditions.first().unwrap();
+            if cond == &TagFiltersAST::NoTagFound {
+                None
+            } else {
+                prune_unusable_filters(conditions.first().unwrap().clone())
+            }
+        }
+        TagFiltersAST::And(conditions) if conditions.len() == 0 => None,
+        TagFiltersAST::And(conditions) if conditions.len() > 1 => {
+            prune_unusable_filters(TagFiltersAST::And(
+                conditions
+                    .into_iter()
+                    .flat_map(|c| prune_unusable_filters(c))
+                    .collect(),
+            ))
+        }
+        TagFiltersAST::Or(conditions) if conditions.len() == 1 => {
+            prune_unusable_filters(conditions.first().unwrap().clone())
+        }
+        TagFiltersAST::Or(conditions) if conditions.len() == 0 => None,
+        TagFiltersAST::Or(conditions) if conditions.len() > 1 => {
+            let pruned_conditions: Vec<TagFiltersAST> = conditions
+                .into_iter()
+                .flat_map(|c| prune_unusable_filters(c))
+                .collect();
+            let is_usefull = !pruned_conditions.contains(&TagFiltersAST::NoTagFound);
+            let condition = TagFiltersAST::Or(pruned_conditions);
+            if is_usefull {
+                Some(condition)
+            } else {
+                None
+            }
+        }
+        ast @ TagFiltersAST::Tag { tag: _ } => Some(ast),
+        TagFiltersAST::Not(ast) => {
+            prune_unusable_filters(*ast).map(|c| TagFiltersAST::Not(Box::new(c)))
+        }
+        TagFiltersAST::NoTagFound => Some(TagFiltersAST::NoTagFound),
+        _ => None,
+    }
+}
+
+fn collect_tag_filters(
+    user_input_ast: &UserInputAst,
+    tag_field_names: &[String],
+) -> Option<TagFiltersAST> {
+    match user_input_ast {
+        UserInputAst::Clause(sub_queries) => {
+            let conditions: Vec<TagFiltersAST> = sub_queries
+                .iter()
+                .flat_map(|(occ, cond)| {
+                    collect_tag_filters(cond, tag_field_names)
+                        .and_then(|c| wrap_with_occur(&occ.unwrap_or(Occur::Should), c))
+                })
+                .collect();
+            Some(TagFiltersAST::Or(conditions))
```

Comment:
I don't think this code is correct.

`[(Must, "a"), (Must, "b")]` here ends up interpreted as  `or([   and(["a"]), and(["b"]) ])`
when it should actually be `and([a. b])`

Reply:
... although to be honest, it may have been the intended spec.As long as the predicate returned is implied by the query, the filter is technically correct. 

If this is the intended behavior, some comments could have been helpful here.

Reply:
@fulmicoton, 
You are right in your diagnosis the `[(Must, "a"), (Must, "b")]` should yield an `and([a. b])`. 
I missed that point. I'll work on it to fix that. 

Thank you for spotting it